### ISSUE TITLE
Compile error: ../../client/model_pool_response_resource.go:34:21: undefined: os

### ISSUE THREAD

**Issue by grokify** [grokify/go-ringcentral-client#26]
# Issue / Symptom

`model_pool_response_resource.go` is missing "os" dependency.

It appears definition `File` gets converted to `*os.File` when it should be be converted to a struct.

```go
package ringcentral

type PoolResponseResource struct {
	RendererId      string     `json:"rendererId,omitempty"`
	MessageId       string     `json:"messageId,omitempty"`
	ExtensionId     string     `json:"extensionId,omitempty"`
	ExtensionNumber string     `json:"extensionNumber,omitempty"`
	FirstName       string     `json:"firstName,omitempty"`
	LastName        string     `json:"lastName,omitempty"`
	Street          string     `json:"street,omitempty"`
	City            string     `json:"city,omitempty"`
	State           string     `json:"state,omitempty"`
	Zip             string     `json:"zip,omitempty"`
	Country         string     `json:"country,omitempty"`
	CountryId       string     `json:"countryId,omitempty"`
	CompanyName     string     `json:"companyName,omitempty"`
	FaxNumber       string     `json:"faxNumber,omitempty"`
	ContactPhone    string     `json:"contactPhone,omitempty"`
	Email           string     `json:"email,omitempty"`
	LanguageCode    string     `json:"languageCode,omitempty"`
	LanguageId      string     `json:"languageId,omitempty"`
	FooterType      float32    `json:"footerType,omitempty"`
	CoverIndex      float32    `json:"coverIndex,omitempty"`
	CoverPageText   string     `json:"coverPageText,omitempty"`
	SourceFiles     []*os.File `json:"sourceFiles,omitempty"`
	MessageServers  []string   `json:"messageServers,omitempty"`
	TouchInterval   int64      `json:"touchInterval,omitempty"`
}
```

The definition includes the following:

```yaml
    PoolResponseResource:
      type: "object"
      properties:
          # ...
          type: "array"
          items:
            $ref: "#/definitions/File"
    File:
      type: "object"
      properties:
        sourceFile:
          type: "string"
        originalFileName:
          type: "string"
```

The `File` struct is still created:

```go
package ringcentral

type File struct {
	SourceFile       string `json:"sourceFile,omitempty"`
	OriginalFileName string `json:"originalFileName,omitempty"`
}
```

Should have `[]File`:

```go
type PoolResponseResource struct {
	SourceFiles     []File `json:"sourceFiles,omitempty"`
}
```

**Comment by grokify** [comment]
# Investigation

## 1) `getSchemaType(Schema p)` function in `AbstractGoCodegen.java`

### Issue

The following values are present when a `#/definition/File` is referenced in an OpenAPI 2.0 spec. It appears, this gets converted to an OpenAPI 3.0 `#/components/schemas/File` path.

| item | value | notes |
|------|--------|--------|
| `String openAPIType = super.getSchemaType(p);` | `File` | |
| `p.get$ref()` | `#/components/schemas/File` | not used |
| return value | `*os.File` | `typeMapping` conversion |

`File` gets converted to `*os.File` due the `typeMapping` entry:

```
typeMapping.put("File", "*os.File");
typeMapping.put("file", "*os.File");
```

```java
    public String getSchemaType(Schema p) {
        String openAPIType = super.getSchemaType(p);
        //...
        if (typeMapping.containsKey(openAPIType)) {
            type = typeMapping.get(openAPIType);
```

`openAPIType` should be kept as `File` instead of being converted to `*os.File` via `typeMapping`.

### A Solution

Return `openAPIType` without transformation when type is from `$ref`:

```java
        String ref = p.get$ref();
        if(ref != null && !ref.isEmpty()) {
            type = openAPIType;
        } else if (typeMapping.containsKey(openAPIType)) {
```

## 2) `getTypeDeclaration(Schema p)` function in `AbstractGoCodegen.java`

### Issue

`getTypeDeclaration` will convert `openAPIType` value of `File` to `*os.File` via `typeMapping`.

```java
        if (typeMapping.containsKey(openAPIType)) {
            typeDeclaration = typeMapping.get(openAPIType);
        }
```

### A Solution

Return `openAPIType` without conversion when it matches OpenAPI v2.0 or v3.0 spec reference as follows.

```java
        String openAPIType = getSchemaType(p);
        String ref = p.get$ref();
        
        if(ref != null && !ref.isEmpty()) {
            String tryRefV2 = "#/definitions/" + openAPIType;
            String tryRefV3 = "#/components/schemas/" + openAPIType;
            if(ref.equals(tryRefV2) || ref.equals(tryRefV3)) {
                return toModelName(openAPIType);
            }
        }
```

## 3) Test in `GoModelTest.java`

The following test is executed with `$ mvn clean package`.

```java
    @Test(description = "convert a property named 'file' model")
    public void filePropertyTest() {
        final Schema model = new Schema()
                .$ref("#/definitions/File");
        final DefaultCodegen codegen = new GoClientCodegen();
        final String ref = model.get$ref();

        Assert.assertEquals(codegen.getSchemaType(model), "File");
        Assert.assertEquals(codegen.getTypeDeclaration(model), "File");
    }
```

## 4) Additional Verification

Verified with:

https://github.com/grokify/api-specs/tree/master/openapicompat

**Comment by grokify** [comment]
Fixed in f434cfbe99ff558cfe04a9d8ebd7cfb90a38c12f
